EIPS/eip-2936.md

```diff
@@ -0,0 +1,66 @@
+---
+eip: 2936
+title: EXTCLEAR Opcode For SELFDESTRUCTed contracts
+author: William Morriss (@wjmelements)
+discussions-to: https://ethereum-magicians.org/t/eip-2936-extclear-for-selfdestruct/4569
+status: Draft
+type: Standards Track
+category: Core
+created: 2020-09-03
+replaces: 2751
+---
+
+`EXTCLEAR` Opcode For `SELFDESTRUCT`ed contracts
+
+## Simple Summary
+Enable new opcode to clear storage for `SELFDESTRUCTED`ed contracts.
+
+## Abstract
+`SELFDESTRUCT` complexity can be reduced by deferring state cleanup.
+
+## Motivation
+`SELFDESTRUCT` (`0xFF`) is unnecessarily complex because it clears an unbounded amount of contract storage.
+It is computationally expensive for nodes to track all of the storage used in every contract in case the contract `SELFDESTRUCT`s.
+Further, contracts can be re-initialized using `CREATE2` (`0xF5`), and then `SLOAD` (`0x54`) prior storage.
+Therefore, several ethereum clients do not clear storage at all, and just check if the contract was initiated since `SSTORE` (`0x55`) during `SLOAD`.
+Nobody expected `SELFDESTRUCT` and `CREATE2` to increase the cost of `SLOAD`.
+Also, bugs in this implementation could split the network.
+
+Instead this defers the time of storage cleanup, and leaves the storage in-place, which reduces the complexity of `SLOAD` and `SELFDESTRUCT`.
+
+This empowers the `CREATE2` reincarnation proxy pattern by retaining storage during upgrade, which would otherwise have to be reset again.
+An atomic reincarnation upgrade could clear a subset of storage during the upgrade, while the contract is destroyed, before reinstating it.
+
+## Specification
+
+After `FORK_BLOCK_NUM`, a new opcode, `EXTCLEAR`, is enabled at `0x5C` to clear storage for `SELFDESTRUCT`ed contracts.
+`EXTCLEAR`:
+* does not push any words onto the stack
+* pops two words off the stack: the destroyed contract address and a storage address
+* if the contract exists, charge the same gas cost as `EXTCODEHASH` (`0x3F`)
+* otherwise, if the storage is zero, charge the same gas as `EXTCODEHASH` plus `SLOAD`
```

Comment:
```suggestion
* otherwise, if the storage is zero, charge the same gas as `EXTCODEHASH` (`0x3f`) plus `SLOAD` (`0x54`)
```
I believe every reference to an opcode in the specification should be in this format.  cc @axic to make sure I have that right.  This includes all of the others below.

Reply:
So, it should be for every opcode, not just the first time it is mentioned?

Reply:
I just double-checked EIP-1, and it appears I was wrong and it only needs to be done the first time.
> If a Core EIP mentions or proposes changes to the EVM (Ethereum Virtual Machine), it should refer to the instructions by their mnemonics and define the opcodes of those mnemonics at least once.

Reply:
Thanks